text/3301-atomic-memcpy.md

```diff
@@ -0,0 +1,227 @@
+It can contain a `T`, but it might be invalid in various ways
+due to concurrent store operations.
+Its interface resembles a mix of the interfaces of `MaybeUninit` and the atomic types.
+
+```rust
+#[repr(transparent)]
+struct AtomicPerByte<T> { inner: UnsafeCell<MaybeUninit<T>> }
+
+unsafe impl<T: Send> Sync for AtomicPerByte<T> {}
+
+impl<T> AtomicPerByte<T> {
+    pub const fn new(value: T) -> Self;
+    pub const fn uninit() -> Self;
+
+    pub fn load(&self, ordering: Ordering) -> MaybeUninit<T>;
+    pub fn store(&self, value: T, ordering: Ordering);
+
+    pub fn load_to(&self, dest: &mut MaybeUninit<T>, ordering: Ordering);
+    pub fn store_from(&self, src: &MaybeUninit<T>, ordering: Ordering);
+
+    pub fn load_to_slice(this: &[Self], dest: &mut [MaybeUninit<T>], ordering: Ordering);
+    pub fn store_from_slice(this: &[Self], src: &[MaybeUninit<T>], ordering: Ordering);
+
+    pub const fn into_inner(self) -> MaybeUninit<T>;
+
+    pub const fn as_ptr(&self) -> *const T;
+    pub const fn as_mut_ptr(&self) -> *mut T;
+
+    pub const fn get_mut(&mut self) -> &mut MaybeUninit<T>;
+    pub const fn get_mut_slice(this: &mut [Self]) -> &mut [MaybeUninit<T>];
+
+    pub const fn from_mut(value: &mut MaybeUninit<T>) -> &mut Self;
+    pub const fn from_mut_slice(slice: &mut [MaybeUninit<T>]) -> &mut [Self];
+}
+
+impl<T> Debug for AtomicPerByte<T>;
+impl<T> From<T> for AtomicPerByte<T>;
+impl<T> From<MaybeUninit<T>> for AtomicPerByte<T>;
+```
+
+Note how the entire interface is safe.
+All potential unsafety is captured by using `MaybeUninit`.
+
+The load functions panic if the `ordering` is not `Relaxed` or `Acquire`.
+The store functions panic if the `ordering` is not `Relaxed` or `Release`.
+The slice functions panic if the slices are not of the same length.
+
+# Drawbacks
+
+- In order for this to be efficient, we need an additional intrinsic hooking into
+  special support in LLVM. (Which LLVM needs to have anyway for C++.)
+
+- It's not immediately obvious this type behaves like a `MaybeUninit`,
+  making it easy to forget to manually drop any values that implement `Drop`.
+
+  This could be solved by requiring `T: Copy`, or by using a better name for this type.
+
+- `MaybeUninit<T>` today isn't as ergonomic as it should be.
+
+  For a simple `Copy` type like `u8` it might be nicer to be able to use types like `&[u8]`
+  rather than `&[MaybeUninit<u8>]`, etc.
+  (But that's a larger problem affecting many other things, like `MaybeUninit`'s interface,
+  `Read::read_buf`, etc. Maybe this should be solved separately.)
+
+# Alternatives
+
+- Instead of a type, this could all be just two functions on raw pointers,
+  such as something like `std::ptr::copy_nonoverlaping_load_atomic_per_byte`.
+
+  This means having to use `UnsafeCell` and more unsafe code wherever this functionality is used.
+
+  It'd be inconsistent with the other atomic operations.
+  We don't have e.g. `std::ptr::load_atomic` that operates on pointers either.
```

Comment:
It would still be convenient to have both the `AtomicPerByte` type and the two functions on raw pointers. This would allow for additional use cases, such as:

* This lets programmers choose if they want to use just a `Seqlock` or combine a `Seqlock` with an `RwLock` to form a hybrid lock ([for example](https://db.in.tum.de/~giceva/papers/damon_latches.pdf?lang=de)).
* The `Seqlock` version counter and the protected data are able to live on separate cache lines, which reduces false sharing while a reader is repeatedly trying to read the contents in a loop.
* The protected data memory layout is unconstrained by the lock, e.g. elements in a continuous array can each be associated with individual locks stored in a separate array, without having to account for this in the data array memory layout, which could otherwise lead to additional unnecessary implicit padding if the elements and the version counter have a different size.

Reply:
Thinking more about this, I believe I got two things mixed up! The `AtomicPerByte` type already allows for all these use cases, although there is another use case for which it is not clear whether it is covered by `AtomicPerByte`:

Does `AtomicPerByte` make any guarantees about whether its `load_to_slice`/`store_from_slice` methods make element-wise atomic copies of `T` or does it do an atomic copy for the entire slice at once? The latter may allow for a more efficient atomic `memcpy` implementation if the slice happens to be word-aligned even though `T` is not (and the slice length is also a multiple of the word-alignment), but then it would be unsound to call methods like `<[AtomicPerByte<T>]>::split_at` that can affect the alignment of the slice if `T` is not word-aligned. If `AtomicPerByte` would promise to only make element-wise atomic copies in this case, then this yields a (very advanced) use case for the two separate pointer methods, which could only be soundly called if the programmer can uphold that there will be no multiple racing atomic `memcpy`s that:

1. operate on an overlapping memory region, and
2. for which one of the copies uses a non-word-aligned offset or a length which is not a multiple of the word size.

otherwise this could lead to mixing non-perfectly-overlapping atomic accesses in the `memcpy` implementation which [might be UB](https://github.com/rust-lang/miri/issues/2303) (assuming that atomic `memcpy` uses the machine word size as the largest granularity for a single instruction, otherwise replace "word size" in this comment with the maximum granularity with which `memcpy` can theoretically operate).

Reply:
would expect `AtomicPerByte` to make per-byte atomic copies, as its name indicates. That means for elements that are larger than 1 byte, this is not even atomic for individual elements.

Reply:
With "element-wise atomic copy" I mean whether the internal implementation of `AtomicPerByte` will perform a separate atomic memcpy method call per element, as opposed to performing a single atomic memcpy method call on the entire slice just once (which may still issue one or more atomic instructions due to making per-byte atomic copies). The difference may matter for the alignment of the pointer that is passed to the atomic `memcpy` method, which may cause the internal implementation of the atomic `memcpy` method to choose a different atomic word size and/or alignment depending on that pointer's alignment, leading to potentially mixing non-perfectly-overlapping atomic accesses in the face of safe methods like `<[AtomicPerByte<T>]>::split_at`, for which the second half of the split slice may have a different alignment in memory and size than the whole slice.

To complete the example of how this could be unsound when `AtomicPerByte` does just a single atomic memcpy method call internally for the entire slice, consider calling `AtomicPerByte::store_from_slice(&whole_slice, ..)` and `AtomicPerByte::load_to_slice(whole_slice.split_at(1).1, ..)`) simultaneously where `whole_slice` is a slice of bytes aligned to the word size or largest SIMD size.

If "*perform a single atomic `memcpy` method call for this slice of `T`*" is not a supported use case of `AtomicPerByte`, then maybe that could be an argument to also include the `atomic_load_per_byte_memcpy` and `atomic_store_per_byte_memcpy` methods in the public API proposed by the RFC? (currently it is phrased to exclude exposing those methods)

Reply:
> With "element-wise atomic copy" I mean whether the internal implementation of AtomicPerByte will perform a separate atomic memcpy method call per element, as opposed to performing a single atomic memcpy method call on the entire slice just once (which may still issue one or more atomic instructions due to making per-byte atomic copies).

I don't think any of that is even defined in the spec; these are all unspecified implementation details.

Reply:
> leading to undefined behavior on x86_64

Despite what the manual says, Intel has a "verbal" guarantee that non-exactly-overlapping atomics is fine: https://github.com/rust-lang/unsafe-code-guidelines/issues/345#issuecomment-1499421708. It may tear across cache lines (which is fine) but will not put the CPU into an undefined state. This situation is unlikely to change given that need to keep C++'s byte-wise memcpy working too.

Reply:
> this has been discussed

That mailing list post was sent on 6 April, all prior discussions were done without this piece of information. The only comment on this RFC on non-exactly-overlapping access since then was yours, so I don't think we can say there's a disposition yet, especially in light of the new information.

In general I don't think "stick to the documentation" is a good approach when dealing with proprietary platforms, otherwise we can't ship anything for macOS at all. If the vendor is receptive, we should nag them to change the documentation like @m-ou-se has done before with MS https://github.com/MicrosoftDocs/sdk-api/pull/447, but even if they don't change it, we shouldn't have our hands tied by it.

In this case, the [C++ proposal](wg21.link/P1478) is committed to allowing multi-byte access optimisations, and they are privy to more information from vendors, so I think sticking to their approach should be safe.